docs/guides/trinity/writing_plugins.rst

```diff
@@ -245,19 +267,69 @@ We may want a plugin to only start running if:
+:meth:`~trinity.extensibility.plugin.BaseIsolatedPlugin.do_start` method to run on an entirely
+separated, new process. In other cases
+:meth:`~trinity.extensibility.plugin.BaseIsolatedPlugin.do_start` will simply run in the same
+process as the plugin manager that the plugin is controlled by.
+
 
 Communication pattern
 ~~~~~~~~~~~~~~~~~~~~~
 
-Coming soon: Spoiler: Plugins can communicate with other parts of the application or even other
-plugins via the central event bus.
+For most plugins to be useful they need to be able to communicate with the rest of the application
+as well as other plugins. In addition to that, this kind of communication needs to work across
+process boundaries as plugins will often operate in independent processes.
 
-Making plugins discoverable
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+To achieve this, Trinity uses the
+`Lahja project <https://github.com/ethereum/lahja>`_, which enables us to operate
+a lightweight event bus that works across processes. An event bus is a software dedicated to the
+transmission of events from a broadcaster to interested parties.
 
-Coming soon.
+This kind of architecture allows for efficient and decoupled communication between different parts
+of Trinity including plugins.
 
-.. warning::
-  **Wait?! This is it? No! This is draft version of the plugin guide as small DEVCON IV gitft.
-  This will turn into a much more detailed guide shortly after the devcon craze is over.**
+For instance, a plugin may be interested to perform some action every time that a new peer connects
+to our node. These kind of events get exposed on the EventBus and hence allow a wide range of
+plugins to make use of them.
+
+For an event to be usable across processes it needs to be pickable and in general should be a
+shallow Data Transfer Object (`DTO <https://en.wikipedia.org/wiki/Data_transfer_object>`_)
+
+Every plugin has access to the event bus via its
+:meth:`~trinity.extensibility.plugin.BasePlugin.event_bus` property and in fact we have already
+used it in the above example to get the current number of connected peers.
+
+.. note::
+  This guide will soon cover communication through the event bus in more detail. For now, the
+  `Lahja documentation <https://github.com/ethereum/lahja/blob/master/README.md>`_ gives us some
+  more information about the available APIs and how to use them.
+
+Distributing plugins
+~~~~~~~~~~~~~~~~~~~~
+
+Of course, plugins are more fun if we can share them and anyone can simply install them through
+``pip``. The good news is, it's not hard at all!
+
+In this guide, we won't go into details about how to create Python packages as this is already
+`covered in the official Python docs <https://packaging.python.org/tutorials/packaging-projects/>`_
+.
+
+Once we have a ``setup.py`` file, all we have to do is to expose our plugin under
+``trinity.plugins`` via the ``entry_points`` section.
+
+.. literalinclude:: ../../../tests/trinity/integration/trinity_test_plugin/setup.py
```

Comment:
@pipermerriam this sucks because we use a different example here than in the rest of the guide which will confuse the reader. Having thought about this, I think it would be better if the example plugin that is used in this guide (`PeerCountReportPlugin`) had a `setup.py` file and then the integration test would simply use that plugin instead of defining its own.

Just, the one thing that bugs me about this plan is that I'm not sure if we want to have things like `setup.py` and the `tar.gz` package within `trinity/plugins/examples`. As in, I'm not sure if that is idiomatic code organization or if that may even come with weird side effects.

I think it would make more sense to have example plugins (or maybe even all plugins?) as a top level directory?

Reply:
> I think it would make more sense to have example plugins (or maybe even all plugins?) as a top level directory?

This was the first thing that came to mind.